log4j-core/src/main/java/org/apache/logging/log4j/core/appender/AsyncAppenderEventForwarder.java

```diff
@@ -0,0 +1,147 @@
+        LOGGER.trace("{} has started.", getName());
+        forwardAll();
+        forwardRemaining();
+    }
+
+    private void forwardAll() {
+        while (!stopped) {
+            LogEvent event;
+            try {
+                event = queue.take();
+            } catch (final InterruptedException ignored) {
+                // Restore the interrupted flag cleared when the exception is caught.
+                interrupt();
+                break;
+            }
+            event.setEndOfBatch(queue.isEmpty());
+            forwardOne(event);
+        }
+        LOGGER.trace("{} has stopped.", getName());
+    }
+
+    private void forwardRemaining() {
+        int eventCount = 0;
+        while (true) {
+            // Note the non-blocking Queue#poll() method!
+            final LogEvent event = queue.poll();
+            if (event == null) {
+                break;
+            }
+            event.setEndOfBatch(queue.isEmpty());
+            forwardOne(event);
+            eventCount++;
+        }
+        LOGGER.trace(
+                "{} has processed the last {} remaining event(s).",
+                getName(), eventCount);
+    }
+
+    void forwardOne(final LogEvent event) {
+
+        // Forward the event to all registered appenders.
+        boolean succeeded = false;
+        // noinspection ForLoopReplaceableByForEach (avoid iterator instantion)
+        for (int appenderIndex = 0; appenderIndex < appenders.size(); appenderIndex++) {
+            final AppenderControl control = appenders.get(appenderIndex);
+            try {
+                control.callAppender(event);
+                succeeded = true;
+            } catch (final Throwable ignored) {
+                // If no appender is successful, the error appender will get it.
+            }
+        }
+
+        // Fallback to the error appender if none has succeeded so far.
+        if (!succeeded && errorAppender != null) {
+            try {
+                errorAppender.callAppender(event);
+            } catch (final Throwable ignored) {
+                // If the error appender also fails, there is nothing further
+                // we can do about it.
+            }
+        }
+
+    }
+
+    void stop(final long timeoutMillis) throws InterruptedException {
+
+        // Mark the completion, if necessary.
+        synchronized (this) {
+            if (!stopped) {
+                stopped = true;
+                LOGGER.trace("{} is signaled to stop.", getName());
+            }
+        }
+
+        // There is a slight chance that the thread is not started yet, wait for
+        // it to run. Otherwise, interrupt+join might block.
+        // noinspection StatementWithEmptyBody
+        while (Thread.State.NEW.equals(getState()));
+
+        // Interrupt the thread. (Note that there is neither a check on
+        // "stopped", nor a synchronization here. It is okay to interrupt
+        // concurrently and/or multiple times.)
+        interrupt();
```

Comment:
So the previous implementation would submit a sentinel event through the queue to wake the background thread instead of relying on thread interruption.
I think we should avoid interrupting the thread if at all possible due to the subtleties of java interruption. Interruption will actually close sockets if any blocking operations are in progress! This means a socket appender may surprisingly fail to deliver final events. I recall some oddities with file i/o as well.

Perhaps we could try to enqueue a sentinel event and fall back to interruption if there's no space in the queue?

Reply:
Makes sense. Will do.

Reply:
Done.